contrib/server/operations/pbench-upload-results.py

```diff
@@ -0,0 +1,270 @@
+#!/usr/bin/env python3.9
+
+"""Upload results tarballs to a Pbench Server
+
+This utility supports bulk migration of results tarballs to a Pbench Server.
+
+It's a standalone utility (with no dependency on any Pbench packages) that can
+be used to send a specific tarball or a potentially large set of tarballs to a
+server. It supports filtering by file modification date, and checkpointing to
+allow restarting in case of a failure. It can be run in a "preview" mode to
+identify the selected tarballs without uploading.
+
+`pbench-upload-results.py https://<server> /srv/pbench/backup` will upload all
+result tarballs nested under the directory `/srv/pbench/backup`. You can use
+the `--since` and `--before` timestamp selectors to include only results files
+modified "since" (>=) and/or "before" (<) the specified datetimes. Specify a
+checkpoint file with `--checkpoint <file>` to record all results successfully
+uploaded; after unexpected termination (e.g., crash or ^C), rerunning the same
+command will skip the already uploaded files.
+
+Uploaded files are marked with metadata consistent with the behavior of the
+0.69-11 passthrough server. If the script is run on a Pbench server, metadata
+will record the version string and SHA1 of the Pbench installation. We also
+record the hostname, and a marker that it was "migrated" by this utility.
+"""
+from argparse import ArgumentParser
+import datetime
+from http import HTTPStatus
+from io import TextIOWrapper
+from pathlib import Path
+import socket
+import sys
+from typing import Optional
+
+from dateutil import parser as date_parser
+import requests
+from requests import Response
+
+
+def get_md5(tarball: Path) -> str:
+    """Read the tarball MD5 from the tarball's companion file
+
+    Args:
+        tarball: Path to a tarball file with a {name}.md5 companion
+
+    Returns:
+        The tarball's MD5 value
+    """
+    md5_file = Path(f"{str(tarball)}.md5")
+    return md5_file.read_text().split()[0]
+
+
+def upload(
+    server: str, token: str, tarball: Path, metadata: Optional[list[str]] = None
+) -> Response:
+    query_parameters = {}
+
+    md5 = get_md5(tarball)
+    query_parameters["access"] = "public"
+    dataset = tarball.name
+    satellite, _ = tarball.parent.name.split("::")
+    meta = [
+        f"global.server.legacy.migrated:{datetime.datetime.now(tz=datetime.timezone.utc):%Y-%m-%dT%H:%M}"
+    ]
```

Comment:
You like specifying your own format better than using `.isoformat()`?
```

    meta = [
        f"global.server.legacy.migrated:{datetime.datetime.now(tz=datetime.timezone.utc).isoformat()}"
    ]
```
(Are you intentionally omitting the fine-granularity time?)

Reply:
I started out with just `"global.server.legacy.migrated:true"` and that felt weird. (And made me think I should have enhanced the query argument `metadata` setter to recognize *types* so that I could get `{"migrated": True}` instead of `{"migrated": "true"}`, e.g. with `global.server.legacy.migrated:true:bool` as I did for filtering.)

I settled on adding the date, even though it's mostly redundant with `dataset.uploaded`, but I didn't want to waste space with the full ISO time, which isn't that interesting. I'd be happy enough to come up with some other semi-meaningful value, but none came to mind. 😆

Reply:
Hmmmm...preserving the original uploaded time might be worth some effort.  Knowing _when_ it was migrated, if that's redundant with the actual upload time from the migration operation is much less interesting than knowing that it was migrated.  🤷

Reply:
Ah ha! Yeah, I like the idea of preserving the "original upload time", which in general ought to correspond with the `st_mtime` of the backup file. (More or less.) Remember we're working from 0.69-11 here, which has no database, and we haven't retained the "archive" tarballs (even if moving them from the ssh intake directory preserves the `mtime`).